libs/json/encoder.go

```diff
@@ -0,0 +1,246 @@
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+	"strconv"
+	"time"
+)
+
+var (
+	timeType            = reflect.TypeOf(time.Time{})
+	jsonMarshalerType   = reflect.TypeOf(new(json.Marshaler)).Elem()
+	jsonUnmarshalerType = reflect.TypeOf(new(json.Unmarshaler)).Elem()
+)
+
+// Marshal marshals the value as JSON, using Amino-compatible JSON encoding (strings for
+// 64-bit numbers, and type wrappers for registered types).
+func Marshal(v interface{}) ([]byte, error) {
+	buf := new(bytes.Buffer)
+	err := encode(buf, v)
+	if err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// MarshalIndent marshals the value as JSON, using the given prefix and indentation.
+func MarshalIndent(v interface{}, prefix, indent string) ([]byte, error) {
+	bz, err := Marshal(v)
+	if err != nil {
+		return nil, err
+	}
+	buf := new(bytes.Buffer)
+	err = json.Indent(buf, bz, prefix, indent)
+	if err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func encode(w io.Writer, v interface{}) error {
+	// Bare nil values can't be reflected, so we must handle them here.
+	if v == nil {
+		return writeStr(w, "null")
+	}
+	rv := reflect.ValueOf(v)
+
+	// If this is a registered type, defer to interface encoder regardless of whether the input is
+	// an interface or a bare value. This retains Amino's behavior, but is inconsistent with
+	// behavior in structs where an interface field will get the type wrapper while a bare value
+	// field will not.
+	if typeRegistry.name(rv.Type()) != "" {
+		return encodeReflectInterface(w, rv)
+	}
+
+	return encodeReflect(w, rv)
+}
+
+func encodeReflect(w io.Writer, rv reflect.Value) error {
+	if !rv.IsValid() {
+		return errors.New("invalid reflect value")
+	}
+
+	// Recursively dereference if pointer.
+	for rv.Kind() == reflect.Ptr {
+		if rv.IsNil() {
+			return writeStr(w, `null`)
+		}
+		rv = rv.Elem()
+	}
+
+	// Convert times to UTC.
+	if rv.Type() == timeType {
+		rv = reflect.ValueOf(rv.Interface().(time.Time).Round(0).UTC())
+	}
+
+	// If the value implements json.Marshaler, defer to stdlib directly. Since we've already
+	// dereferenced, we try implementations with both value receiver and pointer receiver. We must
+	// do this after the time normalization above, and thus after dereferencing.
+	if rv.Type().Implements(jsonMarshalerType) {
+		return encodeStdlib(w, rv.Interface())
+	} else if rv.CanAddr() && rv.Addr().Type().Implements(jsonMarshalerType) {
+		return encodeStdlib(w, rv.Addr().Interface())
+	}
+
+	switch rv.Type().Kind() {
+	// Complex types must be recursively encoded.
+	case reflect.Interface:
+		return encodeReflectInterface(w, rv)
+
+	case reflect.Array, reflect.Slice:
+		return encodeReflectList(w, rv)
+
+	case reflect.Map:
+		return encodeReflectMap(w, rv)
+
+	case reflect.Struct:
+		return encodeReflectStruct(w, rv)
+
+	// 64-bit integers are emitted as strings, to avoid precision problems with e.g.
+	// Javascript which uses 64-bit floats (having 53-bit precision).
```

Comment:
So many good comments ❤️  these should go at least be high lighted in a `doc.go` or maybe even into a markdown document (I'm a fan of package documentation instead of markdown docs because I usually look into code anyways but others prefer specs and markdown docs).